### ISSUE TITLE
test_batch_normalization is failing

### ISSUE THREAD

**Issue by gcp** [intel/mkl-dnn#245]
gcc version 7.3.0
various CPUs (Xeon(R) CPU E3-1240 v3, Ryzen 1700)
Ubuntu 16.04
No MKL
Current git master (72236df2da36d2c73ce1c719c6e9fa04a2f6bdc7)

Doing a simple cmake .. && make && make test produces a test failure:

22/37 Test #22: test_batch_normalization ......................***Failed   45.53 sec

**Comment by emfomenk** [comment]
Hi @gcp,

Could you please provide a little bit more information? Could you please dump the output of the test, so that we can at least see, what exactly fails. Please also set MKLDNN_VERBOSE environment variable to 2 to get a little bit more verbose info.

**Comment by gcp** [comment]
https://sjeng.org/ftp/work/log.txt.gz

relevant part when run with MKLDNN_VERBOSE:
```
[ RUN      ] GoogleNet_Blocked_8/bnrm_test_float.TestsBnrm/2
mkldnn_verbose,create,batch_normalization,jit:avx2,forward_inference,fdata:nChw8c fdiff:undef,flags:0,mb2ic192ih56iw56,0.0490723
mkldnn_verbose,exec,batch_normalization,jit:avx2,forward_inference,fdata:nChw8c fdiff:undef,flags:0,mb2ic192ih56iw56,1.323
mkldnn_verbose,create,batch_normalization,jit:avx2,forward_training,fdata:nChw8c fdiff:undef,flags:0,mb2ic192ih56iw56,0.0959473
mkldnn_verbose,exec,batch_normalization,jit:avx2,forward_training,fdata:nChw8c fdiff:undef,flags:0,mb2ic192ih56iw56,1.10303
mkldnn_verbose,create,batch_normalization,jit:avx2,forward_training,fdata:nChw8c fdiff:undef,flags:1,mb2ic192ih56iw56,0.0629883
mkldnn_verbose,exec,batch_normalization,jit:avx2,forward_training,fdata:nChw8c fdiff:undef,flags:1,mb2ic192ih56iw56,1.59814
mkldnn_verbose,create,batch_normalization,jit:avx2,forward_inference,fdata:nChw8c fdiff:undef,flags:1,mb2ic192ih56iw56,0.0549316
mkldnn_verbose,exec,batch_normalization,jit:avx2,forward_inference,fdata:nChw8c fdiff:undef,flags:1,mb2ic192ih56iw56,1.229
mkldnn_verbose,create,batch_normalization,jit:avx2,forward_inference,fdata:nChw8c fdiff:undef,flags:2,mb2ic192ih56iw56,0.0678711
mkldnn_verbose,exec,batch_normalization,jit:avx2,forward_inference,fdata:nChw8c fdiff:undef,flags:2,mb2ic192ih56iw56,1.17798
mkldnn_verbose,create,batch_normalization,jit:avx2,forward_training,fdata:nChw8c fdiff:undef,flags:2,mb2ic192ih56iw56,0.052002
mkldnn_verbose,exec,batch_normalization,jit:avx2,forward_training,fdata:nChw8c fdiff:undef,flags:2,mb2ic192ih56iw56,1.09985
mkldnn_verbose,create,batch_normalization,jit:avx2,forward_training,fdata:nChw8c fdiff:undef,flags:3,mb2ic192ih56iw56,0.0419922
mkldnn_verbose,exec,batch_normalization,jit:avx2,forward_training,fdata:nChw8c fdiff:undef,flags:3,mb2ic192ih56iw56,1.11719
mkldnn_verbose,create,batch_normalization,jit:avx2,backward_data,fdata:nChw8c fdiff:nChw8c,flags:0,mb2ic192ih56iw56,0.0610352
mkldnn_verbose,exec,batch_normalization,jit:avx2,backward_data,fdata:nChw8c fdiff:nChw8c,flags:0,mb2ic192ih56iw56,1.55908
The difference between (out_diff_src - ref_diff_src) / norm_max and 0. is 0.81878101825714111, which exceeds eps, where
(out_diff_src - ref_diff_src) / norm_max evaluates to -0.81878101825714111,
0. evaluates to 0, and
eps evaluates to 0.62720000743865967.
```

On one of the machines the test passes today (!). Could this be related to uninitialized memory or saving/restoring CPU state? 

I will see if I can make the failure reappear, to exclude the possibility of  having ssh'ed to the wrong machine when I reported it is reproducible on multiple.

**Comment by gcp** [comment]
I managed to reproduce the failure on both systems again. It does not happen if they are idle, but it does happen if there are other machine learning tasks busy (PyTorch+cuDNN).

**Comment by gcp** [comment]
I got a report that this also reproduces on a 56-core AVX-512 machine, without NVIDIA drivers installed.

**Comment by sneves** [comment]
I can reproduce this, on a similar Ubuntu machine as Gian-Carlo. This box is a Skylake with 8 logical cores. Running the following command (with an otherwise idle machine) shows no failures:

    OMP_NUM_THREADS=8 tests/gtests/test_batch_normalization --gtest_filter=GoogleNet_Blocked_8/bnrm_test_float.TestsBnrm/2 --gtest_repeat=-1

However, the command

    OMP_NUM_THREADS=9 tests/gtests/test_batch_normalization --gtest_filter=GoogleNet_Blocked_8/bnrm_test_float.TestsBnrm/2 --gtest_repeat=-1

Fails almost immediately, in the manner described above.

This was also reproduced in a dual-socket Skylake-SP machine with 56 total logical cores. But there the behavior is different, and fewer-than-logical numbers of threads fail. In particular, running the test with 4, 6, 9, 10, 11, 12, 13 threads fails, and after that it seems to stop failing.

**Comment by nastafie** [comment]
Thanks for the details. I've reproduced the issue on Skylake server and am looking into it.

**Comment by beew** [comment]
I am on Ubuntu 16.04 and have the same issue with mkl-dnn-0.14. mkl-dnn is compiled with the small library it downloaded (mklml_lnx_2018.0.3.20180406)

But I also have the intel-mkl lib in my system (2018.2-199)  and openblas installed in my system and switch between them with update-alternatives

Interestingly test_batch_normalization fails system intel-mkl is set to be default blas at run time but it passes if openblas is set to be default blas, moreover it appears to be slightly faster on other tests too.

**Comment by vpirogov** [comment]
Resolved in 1ce0f0bd341cf8942437d50b29a85253197402c4
